// next-i18next/dist/commonjs/types.js
"use strict";

require("core-js/modules/es.object.define-property.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "I18nContext", {
  enumerable: true,
  get: function get() {
    return _reactI18next.I18nContext;
  }
});
Object.defineProperty(exports, "useTranslation", {
  enumerable: true,
  get: function get() {
    return _reactI18next.useTranslation;
  }
});
Object.defineProperty(exports, "Trans", {
  enumerable: true,
  get: function get() {
    return _reactI18next.Trans;
  }
});
Object.defineProperty(exports, "withTranslation", {
  enumerable: true,
  get: function get() {
    return _reactI18next.withTranslation;
  }
});
Object.defineProperty(exports, "Resources", {
  enumerable: true,
  get: function get() {
    return _reactI18next.Resources;
  }
});
Object.defineProperty(exports, "DefaultNamespace", {
  enumerable: true,
  get: function get() {
    return _reactI18next.DefaultNamespace;
  }
});
Object.defineProperty(exports, "appWithTranslation", {
  enumerable: true,
  get: function get() {
    return _.appWithTranslation;
  }
});
Object.defineProperty(exports, "i18n", {
  enumerable: true,
  get: function get() {
    return _.i18n;
  }
});

var _reactI18next = require("react-i18next");

var _ = require("./");
